sdk/core/azure-core/src/main/java/com/azure/core/implementation/AccessTokenCacheImpl.java

```diff
@@ -0,0 +1,169 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.core.implementation;
+
+import com.azure.core.credential.AccessToken;
+import com.azure.core.credential.TokenCredential;
+import com.azure.core.credential.TokenRequestContext;
+import com.azure.core.util.logging.ClientLogger;
+import reactor.core.publisher.Flux;
+import reactor.core.publisher.Mono;
+import reactor.core.publisher.Signal;
+import reactor.core.publisher.Sinks;
+
+import java.time.Duration;
+import java.time.OffsetDateTime;
+import java.util.Objects;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Function;
+import java.util.function.Predicate;
+import java.util.function.Supplier;
+
+/**
+ * A token cache that supports caching a token and refreshing it.
+ */
+public class AccessTokenCacheImpl {
+    // The delay after a refresh to attempt another token refresh
+    private static final Duration REFRESH_DELAY = Duration.ofSeconds(30);
+    // the offset before token expiry to attempt proactive token refresh
+    private static final Duration REFRESH_OFFSET = Duration.ofMinutes(5);
+    private volatile AccessToken cache;
+    private volatile OffsetDateTime nextTokenRefresh = OffsetDateTime.now();
+    private final AtomicReference<Sinks.One<AccessToken>> wip;
```

Comment:
Just trying to get a complete understanding of `wip` and its purpose.

Effectively, from my understanding, there is a background thread which handles requesting refreshed tokens. wip is then fed these tokens which the propagate back to the out going request and authorize it. Correct?

I'm wondering if there is a better way to have this cache offer the last known token than creating a temporary `wip` for potentially every out going request. `Sinks.many` provides the capability to have a sink which replays the last emission for all subscribers. I'm wondering if this may lead to a cleaner implementation.

Reply:
I've updated it to act as the cache for both challenge and non challenge based auth scenarios.
For non challenge based auth scenarios, we can now use the cached token if wip is not available.
For challenge based auth scenarios, we need to wait for the wip as the token request context can change.